core/src/main/kotlin/org/evomaster/core/output/naming/NumberedTestCaseNamingStrategy.kt

```diff
@@ -21,18 +22,51 @@ open class NumberedTestCaseNamingStrategy(
     }
 
     // numbered strategy will not expand the name unless it is using the namingHelper
-    override fun expandName(individual: EvaluatedIndividual<*>, nameTokens: MutableList<String>): String {
+    override fun expandName(individual: EvaluatedIndividual<*>, nameTokens: MutableList<String>, ambiguitySolver: ((Action) -> List<String>)?): String {
         return ""
     }
 
+    override fun resolveAmbiguity(individualToName: MutableMap<EvaluatedIndividual<*>, String>, duplicatedIndividuals: MutableSet<EvaluatedIndividual<*>>) {
+        // do nothing, plain numbered strategy will never have duplicate names
+    }
+
     // kicking off with an empty mutableListOf for each test case to accumulate their own name tokens
     private fun getName(counter: Int, individual: EvaluatedIndividual<*>): String {
         return "test_${counter}${expandName(individual, mutableListOf())}"
     }
 
+    private fun concatName(counter: Int, expandedName: String): String {
+        return "test_${counter}${expandedName}"
+    }
+
     private fun generateNames(individuals: List<EvaluatedIndividual<*>>) : List<TestCase> {
+        val individualToName = mutableMapOf<EvaluatedIndividual<*>, String>()
+        individuals.forEach {
+            // kicking off with an empty mutableListOf for each test case to accumulate their own name tokens
+            ind -> individualToName[ind] = expandName(ind, mutableListOf())
+        }
+
+        val duplicatedNames = getDuplicateNames(individualToName)
+        duplicatedNames.forEach { entry -> resolveAmbiguity(individualToName, entry.value) }
+
         var counter = 0
-        return individuals.map { ind -> TestCase(ind, getName(counter++, ind)) }
+        return individualToName.map { entry -> TestCase(entry.key, concatName(counter++, entry.value)) }
+    }
+
+    private fun getDuplicateNames(individualToName: MutableMap<EvaluatedIndividual<*>, String>): MutableMap<String, MutableSet<EvaluatedIndividual<*>>> {
+        val result = mutableMapOf<String, MutableSet<EvaluatedIndividual<*>>>()
+        individualToName.forEach { entry ->
```

Comment:
this can be simplified with `.map` and `.filter` operations.
as a rule of thumb, should avoid a `forEach` that is filling a mutable list created just for it